[1단계미션] 윤재홍 미션 제출합니다.

yoonjaehong26 opened the pull request:
안녕하세요! 정수영 리뷰어님! 그리디 3 기 프론드앤드 윤재홍입니다.
이번 추석 연휴전에 바쁘실텐데 제 코드를 리뷰해 주셔서 감사드립니다!
아직 리액트에 익숙하지 않지만 열심히 공부하고 있습니다!
***

먼저 미션을 시작하는 저의 상황과 목적에 대해 말씀드리겠습니다.

이번 여름방학에 간단히 클론코딩을 통하여 todo-app 을 만들어 보았지만 js, jsx,html에 대한 기초지식이 없이 시작하여 지금 되돌아 보니 수박 겉핥기 식으로 공부했던 것 같습니다.

이후 바닐라 js와 ejs 로 영단어장 웹을 만들어 보니
동적인 ui구성과 자유로운 페이지이동기능에서 리액트 컴포넌트의 필요성을 느끼게 되었습니다.
[방학동안 재미로 만들어 본 사이트입니다](http://englearnningweb-env.eba-c2pzpg3f.ap-northeast-2.elasticbeanstalk.com)

이러한 이유로 저는 이번 과제에서 컴포넌트 구조구성을 저의 웹과 쿠팡 웹페이지를 보며 리액트의 두가지 원칙, 가장 작은 단위와 재사용을 우선적으로 고려하여 구조를 구성해 보고자 하였습니다.
***
￼
먼저 index.html에서 컴포넌트 단위로 나누어
리액트에 이식하는 작업을 하면서, 태그단위로 나누어져있는 것을 보고 
첫 웹공부가 html책으로 공부했었고 최근 웹크롤링에 대해 알아봤기 때문에 시맨틱 태그 구조를 활용하고자 하였습니다. 
￼￼
<img width="660" height="554" alt="image" src="https://github.com/user-attachments/assets/d4237920-659d-4860-934d-2846d7dfe506" />

따라서 초기 컴포넌트 폴더구성을 위 사진과 같이 시맨틱구조로 구성해보고자 하였습니다.

개발자가 필요한 컴포넌트를 시맨틱구조에서 찾으면 찾기가 쉬울 것이라 생각했지만
 asider 음식점디테일 모달컴퍼넌트로 고민해보니  웹 크롤링시에
 광고와 같이 asider에 위치한다면 홈 음식점 리스트와의 연괸성을 seo가 찾기 어려울 것이라 생각했습니다.
(잘못 알고 있었습니다)

또한 음식점디테일 모달컴퍼넌트가 다른 내 주변의 맛집추천 컴포넌트에서 쓰이게 된다면
컴포넌트의 재사용 시 해당 폴더에서 가져온다는 것이 이상하게 느껴지게 되었습니다.

 알아본 결과 웹 크롤링 엔진이 컴포넌트 폴더구조와 관계없이 
리액트가 최종적으로 렌더링한 html을 기준으로 알아서분석한다는 것을 알게되어 시멘틱 태그구조로 컴포넌트 폴더구성을 하지 않게 되었습니다.
***

만일 홈화면 헤더 컴포넌트를 만든다면 재사용이 불가능해지고 버튼과 내용의 더 작은 단위로 나눌 수 있게되어 이러한 고민에 이어 컴포넌트 단위를 다시 생각해 보았고 레고부품과 연관지어 생각해 보아
우주선 레고 부품과 같은 조립되는 컴포넌트와
우주선과 같은 조립된 컴포넌트로 나누어 생각하게 되어 컴포넌트를 구성해 보았습니다.

***

홈 헤더의 경우 홈 헤더를 만드는 틀과 같은 왼상된 컴포넌트를 폴더에 구성하고 해당 컴포넌트에 조립되는 부품과 같은 컴포넌트를 ui컴포넌트에 저장하는 구조로 만들었습니다.
<img width="1183" height="619" alt="image" src="https://github.com/user-attachments/assets/b4b5cab5-5d14-4a6c-bbd8-de0b575d1b47" />

***
props를 활용하는 것은 2주차에 주로 다루는 것 같아 완성된 컴포넌트로만 구성하고 
이후 css파일을 각 컴포넌트에 맞는  module.css로 배분하였습니다.
***
￼이번 과제를 준비하면서 리액트 컴포넌트의 구조에 대해 왜 '2가지 원칙, 가장 작은 단위와 재사용'을 
이해해 보려 노력해봤는데  기본 문법만 공부하고 "일단 코드를 짜보자!"고 생각해 위와 같은 과정으로 생각하였는데
이미 구조를 짜고 난 뒤에 잘못된 구조로 설계한 것 같습니다..

위 pr은 제가 과제를 하면서 노션에 적었던 과정들을 그대로 작성해 보았는데 다른 코드를 짤 때에도 이와 같이 
여러번 길을 벗어난 적이 많아서
프로그래밍을 설계하고 공부하여 코드를 작성해 나가는 방식이 아직 미흡한 점이 많이 느껴져
정수영 리뷰어님께서는 어떻게 코드를 설계해 나가시는지  궁금합니다!
***
또한 제가 생각한 컴포넌트 구조가 적절한지 궁금합니다!
앞서 말씀드린 대로 2페이지의 제 사이트와 여러개의 페이지로 구성된 쿠팡 웹페이지를 생각하며 구성해 보았는데
지금은 컴포넌트가 하나의 구역에서만 쓰이지만 여러곳의 구역에서 컴포넌트가 쓰이게 된는 등 어떤점을 더 고려하여 
컴포넌트를 설계하는 지 궁금합니다.

<img width="940" height="990" alt="image" src="https://github.com/user-attachments/assets/f2ffd088-da0a-484e-95c2-8c797e63b16a" />
<img width="973" height="978" alt="image" src="https://github.com/user-attachments/assets/e0761d2a-83b8-49d0-ab69-ca6adc136169" />

Commit:
refactor(/src):index.html dom요소 리액트 컴포턴트로 재구성

css는 전역 App.css로 임시적용함.이미지 파일 연동 안됨

Commit:
refactor(HomeHeader.jsx,RestrauantList.jsx):이미지 파일 컴포넌트에 연동

Commit:
refactor(/src):style.css파일 module.css로 분리,문법에 맞게 수정

Commit:
refactor(/src,eslintrc.cjs,package-lock.json,package.json):eslint설치 밎 적용

1.문자열 큰따움표샤용 => 작은따움표 사용
2.줄바꿈 에어비앤비 컨벤션 적용
3.dom태그 문법 에어비앤비 컨벤션 적용

Commit:
refactor(/src):코드마지막줄 개행 추가

Commit:
fix(App.jsx, App.css):App.module.css와 중복되어 적용되는 css파일 삭제

Commit:
fix(HomeHeader.jsx):HomeHeader.module.css가 적용되지 않은 부분 수정

Commit:
refactor(eslintrc.cjs):코드마지막줄 개행 추가

Commit:
fix(App.jsx, RestaurantList.jsx):파일명, 컴포넌트명이름 불일치 수정

suyoungj commented on src/App.jsx:
```diff
@@ -1,7 +1,27 @@
-import "./App.css";
+import styles from './App.module.css';
```
`.module.css`를 사용하신 이유가 궁금합니다.
`.css`와는 어떤 차이점이 있나요??

yoonjaehong26 replied:
<img width="458" height="54" alt="image" src="https://github.com/user-attachments/assets/db57df7a-7160-47f4-8755-88e97aa1d953" />

.module.css의 가장 큰 특징은 전역적으로 적용되는 css와 달리 import된 파일 내에서만 적용이 된다는 점입니다.
위 사진과 같이 .module.css의 각 선택자에 고유한 값이 부여되고 해당 컴포넌트에서만 적용됩니다.

컴포넌트 단위로 작성되는 리액트에서 개발자가 보다 더 작은 단위에서 css를 관리할 수 있게되어
선택자 중복에 대한 문제를 신경쓰지 않아도 되어 유지보수에 더 유용한 것으로 알고 있습니다!

suyoungj commented on src/components/Header/HomeHeader.jsx:
```diff
@@ -0,0 +1,14 @@
+  return (
+    <header className={styles.gnb}>
+      <h1 className={`${styles.gnb__title} text-title`}>점심 뭐 먹지</h1>
+      <button type="button" className={styles.gnb__button} aria-label="음식점 추가">
```
`aria-label`은 직접 작성하신걸까요?
`<button>`과 같은 엘리먼트에 왜 `aria-label`을 추가해야 할까요?

yoonjaehong26 replied:
```JS
//index.html
 <!-- GNB -->
  <header class="gnb">
    <h1 class="gnb__title text-title">점심 뭐 먹지</h1>
    <button type="button" class="gnb__button" aria-label="음식점 추가">
      <img src="./add-button.png" alt="음식점 추가">
    </button>
  </header>
```
위 index.html에서 컴포넌트 단위로 분리하여 바로 가져온 부분이라 제가 추가한 내용은 아닙니다!
***

 화면을 볼 수 없는 시각장애우분들이 웹을 이용하실 때  스크린 리더를 사용하여 웹의 텍스트를 변환하여 음성으로 듣고 사용하시는데
이때 버튼이나 이미지처럼 텍스트로 이루어 지지 않은 요소들을 텍스트로 알려주기 위하여 있는 문법입니다.

스크린 리더는 웹 내 dom태그 구조를 바탕으로 사용자에게 정보를 제공하므로 버튼이나 입력창과 같은 경우 태그종류와 태그 상태를 
나타내기도 하고 ` area-label`이나   `<div role="태그에 대한 설명">`, `<ul>`, `<ol>`,` <li>`,`<h1~h6>` `<art=''>`와 같은 
태그 사용등으로 스크린 리더에게 요소의 맥락과 조작상태를 나타낼 수 있습니다.

따라서 seo나 시매틱구조 뿐 아니라 스크린 리더를 위해 웹 구조의 맥락을 고려하며 구성해야 할 필요성이 있는 것 같습니다!

suyoungj commented on src/components/Header/HomeHeader.jsx:
```diff
@@ -0,0 +1,14 @@
+ 
+function HomeHeader() {
+  return (
+    <header className={styles.gnb}>
```
`<header> `가 `App.jsx`와 더불어 중복되어 사용되고 있네요.
컴포넌트 네이밍에 `Header`를 포함하셔서, `<header>` 태그도 이 컴포넌트에서 사용하는게 덜 헷갈릴 것 같아요.

yoonjaehong26 replied:
컴포넌트로 분리하는 작업중에  App.jsx와 HomeHeader.jsx에서 중복되어 사용된 것을 놓쳤습니다.
App.jsx 에서 Fragment를 사용하여 중복문제를 해결하겠습니다. 감사합니다!

96a5b509d7fc818ec3cf576a1875853a4b037545

suyoungj commented on src/components/Main/RestaurantCategoryFilter.module.css:
```diff
@@ -0,0 +1,22 @@
+  margin-top: 24px;
+}
+
+.restaurantFilterContainer select {
```
A) `.restaurantFilterContainer select`
B) `.restaurantFilterContainer > select`

A, B는 어떤 차이점이 있을까요?

yoonjaehong26 replied:
A)와B)는 자식 엘리먼트에 대한 스타일 적용 문법입니다.  

` <section className={styles.restaurantFilterContainer}>` 의 자손인 ` <select name="category">`에 적용되는 문법입니다.

다만 B)의 경우`는 .restaurantFilterContainer`의 직계자식, 즉 부모 엘리먼트의 바로 아래에 속한 자식만 해당이 되지만
A)의 경우` .restaurantFilterContainer `직계 자식인 `select`뿐 아니라 `select`의 자식또한 범위에 포함이 됩니다.

suyoungj commented on src/components/Main/RestaurantList.jsx:
```diff
@@ -0,0 +1,77 @@
+import styles from "./RestrauantList.module.css";
+    
+function RestrauantList() {
```
gemini 리뷰대로 오타가 있네요.
다소 무겁지만, [Code spell checker](https://open-vsx.org/extension/streetsidesoftware/code-spell-checker) 익스텐션을 활용하면 오타를 많이 줄일 수 있어 추천드립니다!

yoonjaehong26 replied:
네 확장 프로그램 설치해보니 오타를 알려주는 기능이 있네요
오타 수정하였고 다른 오탈자 있는지 확인하였습니다 감사합니다!
145a45ac3871f72254963b0e162b70a076e09416

suyoungj commented on src/components/Main/RestaurantList.jsx:
```diff
@@ -0,0 +1,77 @@
+
+        <li className={styles.restaurant}>
+          <div className={styles.restaurantCategory}>
+            <img src="templates\category-chinese.png" alt="중식" className={styles.categoryIcon}/>
```
`alt`를 직접 작성하신걸까요? 
`<img>` 태그에 `alt`는 왜 필요할까요?

yoonjaehong26 replied:
```JS
//index.html
 <header class="gnb">
    <h1 class="gnb__title text-title">점심 뭐 먹지</h1>
    <button type="button" class="gnb__button" aria-label="음식점 추가">
      <img src="./add-button.png" alt="음식점 추가">
    </button>
  </header>
```
두번째 질문에서와 같이 index.html에서 그대로 컴포넌트로 가져와서 제가 직접 추가한 것은 아닙니다.
***

alt의 사용 이유에는 크게 3가지가 있는 것으로 알고 있는데
첫번째는 이미지가 웹에서 렌더링 되지 않을 때 대체하여 텍스트로 보여지는 역할을 합니다.
<img width="406" height="514" alt="image" src="https://github.com/user-attachments/assets/948ee5c5-2a14-48b1-b2fa-62816817c70f" />
<출처:web.[dev](https://web.dev/learn/performance/understanding-the-critical-path?hl=ko)>
이미지가 아직 렌더링 되지 않은 가장 큰 이유는, 첫 html 파싱과정에서 이미지 파일을 찾게 되면 외부 서버나 public폴더에서 
이미지를 요청하게 되는데 pain단계에서 아직 이미지가 준비되지 않으면 이때 alt의 값이 보여지게 됩니다.
***
두번째로는 앞선 질문에서 말씀드린 스크린 리더에 이미지에 대한 정보를 넘겨주는 역할을 할 수 있습니다.
따라서 이미지에 대한 전체적인 맥락이나 다른 domcument와의 관계를 나타내면 좋습니다.
***
세번째로는 seo(검색엔진 최적화)를 통해 웹을 탐색할 때 이미지를 읽지 못하기 때문에 이를 대체하여 이미지에 대한 정보를 나타내는 역할을 합니다.
위 스크린리더와 같이 자세한 정보나 맥락을 기제하면 좋지만 과도하게 많은 키워드를 사용 시 핵심 키워드가 분산되는 문제가 생길 수 있습니다.
***
따라서 다양한 관점을 고려할 때 이미지에 alt태그를 사용하는 것이 단순한 화면구성 이외의 여러 기능을 가지기 때문에 
사용하는것이 권장되는 것 같습니다.

suyoungj commented on src/components/Aside/RestaurantModal.module.css:
```diff
@@ -0,0 +1,125 @@
+  color: var(--grey-300);
+}
+
+input[name="name"],
```
`input` vs. `input[name="name"]`
어떤 선택자가 우선순위가 더 높을까요?

yoonjaehong26 replied:
css 적용 우선순위로 알아볼 수 있으며
`!important`  => `인라인 스타일`  => `아이디 선택자` =>
`클래스 / 어트리뷰트 / 가상 선택자`  => `태그 선택자`  =>` 전체 선택자` =>
`상위 요소에 의해 상속된 속성`와 같은 우선순위가 적용되어 
태그선택자인 `input` 보다 어트리뷰트문법이 사용된  `input[name="name"]`이 더 우선적으로 적용되게 됩니다.

더 찾아보니 css 명시도라는 개념이 있어 알아봤습니다.
<img width="727" height="349" alt="image" src="https://github.com/user-attachments/assets/0de8786d-7766-4844-b18f-49e316149b11" />
위 표에 따라 `input`는 0001의 명시도값을 가지게 되고  `input[name="name"]`는 0010의 명시도 값을 가지게 되어
둘다 클라스를 가지거나  동일한 인라인 스타일 내에서도  `input[name="name"]`이 적용되게 됩니다.

suyoungj replied:
css 스타일이 예상대로 적용되지 않을 때는 명시도 문제일 가능성이 꽤 높아요.
관련해서 학습에 도움이 될 만한 자료 공유드립니다 🙂

- [MDN Specificity](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_cascade/Specificity)
- [SpeciFISHity](https://specifishity.com)

suyoungj reviewed (changes requested):
재홍님 안녕하세요. 3주차 리뷰를 맡게 된 정수영입니다.

리뷰에 앞서 마감 기한 관련해서 한 가지 말씀드리자면,
상황이 헷갈릴 수 있다는 점은 이해하지만 이런 맥락을 미리 리뷰어에게 공유해주시면 일정 이해에 도움이 돼요.
특히 프론트엔드는 개발 일정상 기획, 디자인, 백엔드 API 개발이 모두 필요한 불리한 위치에 놓여 있어 일정 관리는 조금 더 신경써주셨으면 좋겠어요 🙂

전반적으로 템플릿을 적정 단위 컴포넌트로 쪼개서 렌더링하는 요구사항으로 보여서, 코드 상으로는 리뷰할 점이 별로 없었어요. 컴포넌트 각각의 단위도 적절하게 잘 나눠주신것 같아요 👍

### 질문에 대한 답변
Q1)
> 위 pr은 제가 과제를 하면서 노션에 적었던 과정들을 그대로 작성해 보았는데 다른 코드를 짤 때에도 이와 같이 여러번 길을 벗어난 적이 많아서
프로그래밍을 설계하고 공부하여 코드를 작성해 나가는 방식이 아직 미흡한 점이 많이 느껴져
정수영 리뷰어님께서는 어떻게 코드를 설계해 나가시는지 궁금합니다!

A1)
꼭 필요한 과정을 거치고 계신 거라고 생각해요. 많이 만들어보고, 부족한 점을 개선하고, 잘 된 점은 체화해서 바로 적용할 수 있도록 꾸준히 학습하시다 보면 감이 잡히실거라 생각해요.

저는 문제가 복잡할 때는 먼저 작은 단위로 나누고, 머릿속에 잘 떠오르지 않으면 노트에 적어보면서 개략적으로 설계한 뒤에 코드를 작성하는 편이에요. 
설계는 매번 어렵게 느껴져서 😅, 오래 공들여 완벽하게 설계하기보단 빠르게 구현하고 부족한 점을 바로바로 개선하는 방식을 선호해요.

---

Q2)
> 또한 제가 생각한 컴포넌트 구조가 적절한지 궁금합니다!
앞서 말씀드린 대로 2페이지의 제 사이트와 여러개의 페이지로 구성된 쿠팡 웹페이지를 생각하며 구성해 보았는데 지금은 컴포넌트가 하나의 구역에서만 쓰이지만 여러곳의 구역에서 컴포넌트가 쓰이게 된는 등 어떤점을 더 고려하여 컴포넌트를 설계하는 지 궁금합니다.

A2)
현재 규모의 애플리케이션에서는 적절해 보여요. 
말씀주신 것처럼 **지금은 가장 작은 단위, 재사용성만 고려해서 잘 적용해 주시는 것으로 충분**합니다.
- 동일한 UI 요소가 반복되는 경우 (재사용성 고려)
- 하나의 컴포넌트가 너무 크고 여러 역할을 담당하는 경우 (하나의 역할을 담당하는 작은 단위의 컴포넌트로 분리 및 조합)

Commit:
fix(App.jsx):App.jsx와 HomeHeade.jsx에서 <header>가 중복되던 부분 제거

Commit:
refactor(RestaurantList.jsx, RestaurantList.module.css):파일명 및 관련 네이밍 오타 수정

yoonjaehong26 reviewed (commented):
정수영 리뷰어님 1주차 리뷰 반영 및 답변 완료하였습니다.
이번 과제를 준비하면서 생각하였던 과정을 pr에 나타내 보려고 하였는데 제가 다시 봐도 
읽기에 불편한 글이였고 불필요한 부분이 많았던 것 같습니다.

이번 주말동안 과제에 대한 준비와 계획에 대한 부족한 점들을 고민하였고 말씀해 주신 것처럼
일정관리에 대하여 철저히 준비해 나가겠습니다.

코드 설계방법에 대해, 아직 100줄 이상 규모의 코드작성에 익숙하지 않지만  작은 단위로 나누어서
조금씩 발전해가는 방식으로 익숙해져 보겠습니다!

리뷰 해주신 정수영 리뷰어님께 정말 감사드립니다.

Commit:
refactor():에어비앤비 eslint규칙에 따른 코드 수정

1.리액트 17.0버전 이후 jsx확장자 더이상 사용하지 않아서 삭제
2. 새로고침과 같은 의도치 않은 작동을 막기 위하여 버튼 태그에 type="button추가
3. 줄바꿈및 depth길이 통일화
4. 큰따움표("")대신 작은따움표 사용('')
eslintrc 규칙 수정:
1.'linebreak-style': 0: 맥과 윈도우 환경에서 줄바꿈 스타일차이로 인한 임시로 규칙 무시
2."react/react-in-jsx-scope": "off" : 리액트 17.0버전 이후 jsx확장자 더이상 사용하지 않아 규칙 무시

yoonjaehong26 commented on src/components/Main/RestaurantList.jsx:
```diff
@@ -0,0 +1,77 @@
+
+        <li className={styles.restaurant}>
+          <div className={styles.restaurantCategory}>
+            <img src="templates\category-korean.png" alt="한식" className={styles.categoryIcon}/>
```
리뷰 반영하여 수정 완료하였습니다.
40b4a8d02171f71de18a79e77f39ea9f272cf298

yoonjaehong26 commented on src/components/Header/HomeHeader.jsx:
```diff
@@ -0,0 +1,14 @@
+ 
+function HomeHeader() {
+  return (
+    <header className={styles.gnb}>
+      <h1 className={`${styles.gnb__title} text-title`}>점심 뭐 먹지</h1>
+      <button type="button" className={styles.gnb__button} aria-label="음식점 추가">
+        <img src="templates\add-button.png" alt="음식점 추가" />
+      </button>
+    </header>
```
리뷰 반영 및 코드수정 완료하였습니다.
App.jsx에서  </header>태그대신 <></>프래그먼트 사용하여 코드 수정 완료
https://github.com/cho-log/self-paced-react/commit/96a5b509d7fc818ec3cf576a1875853a4b037545

yoonjaehong26 commented on src/components/Main/RestaurantList.jsx:
```diff
@@ -0,0 +1,77 @@
+import styles from "./RestrauantList.module.css";
+    
+function RestrauantList() {
```
오타 수정완료
https://github.com/cho-log/self-paced-react/commit/145a45ac3871f72254963b0e162b70a076e09416

yoonjaehong26 commented on src/App.jsx:
```diff
@@ -1,7 +1,27 @@
+        <RestrauantList />
+      </main>
 
+      <article>
+        <RestaurantDetailModal />
+        <AddRestaurantModal />
+      </article>
```
중복된 <header>,<article>대신 <></>으로 대체하여 코드 수정 완료
6266a8fa530415ed246ee2c32468e2cd7111e69d
